### DISCUSSION TITLE
0.13.1 Columns Selection DSL, KDocs, Table Rendering, and Many Fixes!

### DISCUSSION THREAD

**Discussion by Jolanrensen** [Kotlin/dataframe#633] [Announcements]
We just released v0.13.1!

Documentation/Readme might not be up-to-date yet. However, feel free to test it in your project and let us know if something does not work as expected!

To sum up the most significant changes: 
- We finally merged the new [ColumnsSelection DSL](https://github.com/Kotlin/dataframe/pull/372) (documentation will [come](https://github.com/Kotlin/dataframe/issues/570) too). This comes with KDocs everywhere, previously missing overloads, and clearer function names.
- We're continuously improving support for [Kotlin Notebook](https://plugins.jetbrains.com/plugin/16340-kotlin-notebook) by creating and improving a native table component for DataFrames, by @ermolenkodev 
- (Nested) table output now looks better for notebooks on Github (for [example](https://github.com/Kotlin/dataframe/blob/0.13.0-RC1/examples/notebooks/github/github.ipynb))
- Improvements to Arrow reading (thanks, @fb64 and @Kopilov!)
- Many other fixes and version bumps, see below!

Check it out on [Maven Central](https://central.sonatype.com/artifact/org.jetbrains.kotlinx/dataframe/0.13.1)

## Features
* ColumnsSelection DSL overhaul by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/372
* Add static rendering for GitHub by @ileasile in https://github.com/Kotlin/dataframe/pull/480
* Static notebook fixes and examples by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/511
* Better static tables by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/519
* Added ! to DataColumnArithmetics by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/531
* Add read of Arrow TimeStamp without timezone as LocalDatetime #515 by @fb64 in https://github.com/Kotlin/dataframe/pull/516
* Add readArrowReader method to allow loading a dataframe from an ArrowReader by @fb64 in https://github.com/Kotlin/dataframe/pull/529
* Excel add new sheet without overwriting the file by @LeandroC89 in https://github.com/Kotlin/dataframe/pull/157
* make `forEach` inline to enable composable and suspend calls in lambda by @koperagen in https://github.com/Kotlin/dataframe/pull/572
* Add option to prevent Gradle plugin from adding dependency on KSP by @koperagen in https://github.com/Kotlin/dataframe/pull/571
* add infer types parameter in DSL functions by @koperagen in https://github.com/Kotlin/dataframe/pull/579
* Reading Arrow NullVector by @Kopilov in https://github.com/Kotlin/dataframe/pull/550

## Fixes
* Fix performance problem in `rename` implementation by @nikitinas in https://github.com/Kotlin/dataframe/pull/532
* Ported the fix for JDBC integration from the 0.12.1 branch by @zaleslaw in https://github.com/Kotlin/dataframe/pull/538
* Fix reading Excel file that has cells with formulas that return string by @jmrsnt in https://github.com/Kotlin/dataframe/pull/484
* Catching Error too in `isOpenApiStr` by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/490
* Turning off the feedback widget in documentation as it's non-functional by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/491
* Fixes related to Kotlin Notebook plugin integration by @ermolenkodev in https://github.com/Kotlin/dataframe/pull/501
* remove klaxon usage outside json.kt by @koperagen in https://github.com/Kotlin/dataframe/pull/506
* use range indexing to get rows subset by @koperagen in https://github.com/Kotlin/dataframe/pull/514
* Linter for all modules by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/521
* add analytics initialization script by @koperagen in https://github.com/Kotlin/dataframe/pull/524
* Fix toDataFrame for value classes by @koperagen in https://github.com/Kotlin/dataframe/pull/542
* make sure to adjust nullability in order to fix platform types by @koperagen in https://github.com/Kotlin/dataframe/pull/552
* Fixes Api Guru examples by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/559
* generate "nullable properties" only where needed by @koperagen in https://github.com/Kotlin/dataframe/pull/597
* Updating maven publication developers list by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/611
* Fix for #573: Change serialization format for rendering in IntelliJ IDEA  by @ermolenkodev in https://github.com/Kotlin/dataframe/pull/574
* Fix for mac-os permissions pre-commit by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/616
* prevent Klaxon from using default reflective serializer by @koperagen in https://github.com/Kotlin/dataframe/pull/621

## Docs and Examples
* update groupBy documentation by @koperagen in https://github.com/Kotlin/dataframe/pull/465
* Docs updates by @koperagen in https://github.com/Kotlin/dataframe/pull/545
* Update of the documentation for the 0.12.1 release by @zaleslaw in https://github.com/Kotlin/dataframe/pull/556
* Documentation updates by @koperagen in https://github.com/Kotlin/dataframe/pull/560
* migrate Titanic.ipynb to Kandy by @koperagen in https://github.com/Kotlin/dataframe/pull/548
* fix broken links by @koperagen in https://github.com/Kotlin/dataframe/pull/626
* Checking notebooks for 0.13.1 by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/625
* Fixes in resize-iframes.js by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/618

## Version Updates
* Update simple-git version to 2.0.3 by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/535
* Updating docProcessor to 0.3.3 by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/600
* Ksp fix by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/592
* Version bumps by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/590
* Version bumping Dokka to 1.9.10 by @Jolanrensen in https://github.com/Kotlin/dataframe/pull/599
* Updated MySQL and PostgreSQL versions in build config by @zaleslaw in https://github.com/Kotlin/dataframe/pull/609

## New Contributors
* @jmrsnt made their first contribution in https://github.com/Kotlin/dataframe/pull/484
* @fb64 made their first contribution in https://github.com/Kotlin/dataframe/pull/516
* @LeandroC89 made their first contribution in https://github.com/Kotlin/dataframe/pull/157

## Known Issues
*  KDoc rendering looks odd on 2024 Nightly IntelliJ by @Jolanrensen in https://github.com/Kotlin/dataframe/issues/585
   * This is being worked on in [IntelliJ](https://youtrack.jetbrains.com/issue/IDEA-348571/Markdown-KDoc-rendering-errors), [JetBrains Markdown](https://github.com/JetBrains/markdown/pull/150), and [here](https://github.com/Kotlin/dataframe/pull/613).

**Full Changelog**: https://github.com/Kotlin/dataframe/compare/v0.12.1...v0.13.1

<hr /><em>This discussion was created from the release <a href='https://github.com/Kotlin/dataframe/releases/tag/v0.13.1'>0.13.1 Columns Selection DSL, KDocs, Table Rendering, and Many Fixes!</a>.</em>

**Comment by pacher** [comment]
Was always too shy to ask, is native table component for DataFrames closed-source? Is it possible to use it in own projects without notebooks?

**Reply by Kopilov** [reply]
@pacher DataFrame library is absolutely open source and free to bu used in any projects (Apache-2.0 license)

I have made two projects with this library, both are Gradle-built backend or just CLI

**Reply by pacher** [reply]
Hey @Kopilov Thanks!
The question is not about kdf, but about the new Swing (what they call native) component rendering tables in the kotlin notebooks plugin for IDEA. As opposed to html table rendering which was used before and which is indeed part of the open source dataframe library.

**Reply by ermolenkodev** [reply]
@pacher Yes, the UI component is currently closed-source and tightly integrated with the IntelliJ IDEA Ultimate SDK.

**Reply by ermolenkodev** [reply]
Yes, indeed it is native to the IDE, not to the OS. :)
